mycore-user2/src/main/java/org/mycore/user2/MCRUser.java

```diff
@@ -710,27 +714,16 @@ private void setRoles(MCRRole[] roles) {
             .forEach(this::assignRole);
     }
 
-    @Transient
-    @XmlElementWrapper(name = "attributes")
-    @XmlElement(name = "attribute")
-    private MapEntry[] getAttributesMap() {
-        if (attributes == null) {
-            return null;
-        }
-        ArrayList<MapEntry> list = new ArrayList<>(attributes.size());
-        for (Entry<String, String> entry : attributes.entrySet()) {
-            MapEntry mapEntry = new MapEntry();
-            mapEntry.name = entry.getKey();
-            mapEntry.value = entry.getValue();
-            list.add(mapEntry);
-        }
-        return list.toArray(new MapEntry[list.size()]);
-    }
-
-    @SuppressWarnings("unused")
-    private void setAttributesMap(MapEntry[] entries) {
-        for (MapEntry entry : entries) {
-            attributes.put(entry.name, entry.value);
+    public void setUserAttribute(String name, String value) {
+        Optional<MCRUserAttribute> anyMatch = getAttributes().stream()
+            .filter(a -> a.getName().equals(Objects.requireNonNull(name)))
+            .findAny();
+        if (anyMatch.isPresent()) {
+            MCRUserAttribute attr = anyMatch.get();
+            attr.setValue(value);
+            getAttributes().removeIf(a -> a.getName().equals(name) && a != attr);
+        } else {
+            getAttributes().add(new MCRUserAttribute(name, value));
```

Comment:
If the order of attributes is not relevant (as it was in the previous map), it could just be:
 getAttributes().removeIf(a -> a.getName().equals(name));
 getAttributes().add(new MCRUserAttribute(name, value));

Reply:
saves a database instruction in case an attribute is present (UPDATE vs. DELETE & INSERT) at the cost of more local instructions